client/components/webpack-build-monitor/index.tsx

```diff
@@ -0,0 +1,153 @@
+/**
+ * @see https://webpack.js.org/api/module-variables/#__webpack_hash__-webpack-specific
+ */
+declare const __webpack_hash__: string;
+
+/**
+ * External dependencies
+ */
+import React, { FunctionComponent, useEffect, useState } from 'react';
+import classNames from 'classnames';
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import Spinner from 'components/spinner';
+import { IntervalHandle, TimestampMS } from 'client/types';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+const CONNECTION_TIMEOUT = 20 * 1000;
+const debug = debugFactory( 'calypso:webpack-build-monitor' );
+
+enum BuildState {
+	UNKNOWN,
+	IDLE,
+	BUILDING,
+	BUILT_WITH_WARNINGS,
+	ERROR,
+}
+
+const WebpackBuildMonitor: FunctionComponent = () => {
+	const [ lastHash, setLastHash ] = useState( __webpack_hash__ );
+	const [ isConnected, setIsConnected ] = useState( false );
+	const [ buildState, setBuildState ] = useState( BuildState.UNKNOWN );
+
+	useEffect( () => {
+		if ( typeof EventSource === 'undefined' ) {
```

Comment:
I prefer to keep my `useEffect` functions small by extracting as much as possible to a separate function:
```js
useEffect( () => connectToWebpack( { setLastHash, setIsConnected, setBuildState } ), [] );
```

`connectToWebpack` can be outside the `WebpackBuildComponent` and they have a clear interface between them.

Reply:
By the way, did you know that you don't need to implement the HMR client yourself, but can use this?
```js
import { subscribeAll } from 'webpack-hot-middleware/client';

subscribeAll( message => {
  // handle parsed JSON message
  // the heartbeat emoji messages have been filtered out,
  // you get everything else as is
} );
```

Reply:
> By the way, did you know that you don't need to implement the HMR client yourself, but can use this?
> 
> ```js
> import { subscribeAll } from 'webpack-hot-middleware/client';
> 
> subscribeAll( message => {
>   // handle parsed JSON message
>   // the heartbeat emoji messages have been filtered out,
>   // you get everything else as is
> } );
> ```

I started with that, found that I didn't have access to the hot update results to know whether reload was needed, and thought I might be able to get that by listening to the EventSource directly.

That didn't work out, but considering we'd need to handle applying updates in another way to know whether the application is up to date or not, I left this implementation.

Reply:
> I started with that, found that I didn't have access to the hot update results to know whether reload was needed

The are two ways how to resolve this:

After implementing the `EventSource` client yourself, you're just a few lines away from having a complete HMR client. The only thing that's missing is performing the hot update:
```js
await module.hot.check(); // download the updated manifest and chunks
await module.hot.apply( applyOptions ); // apply the downloaded update
```
or in a shortcut form:
```js
await module.hot.check( applyOptions );
```
`module` is any CommonJS module, as in `module.exports`. That means the client must be a CommonJS module. I don't know any way how to grab `module.hot` in an ES module. (it's a webpack-specific module API)

Create React App doesn't use any existing HMR client either, but [implements its own one](https://github.com/facebook/create-react-app/blob/master/packages/react-dev-utils/webpackHotDevClient.js).

Another way is to keep `webpack-hot-middleware/client` as is, but subscribe to its status updates:
```js
module.hot.addStatusHandler( listener );
module.hot.removeStatusHandler( listener );
```
The status cycles through a `idle` -> `check` -> `ready` -> `apply` -> `idle` loop.

Reply:
> After implementing the EventSource client yourself, you're just a few lines away from having a complete HMR client.

I noticed this, but didn't finish the implementation because a React component seems like the wrong fit for managing the updates. I'd like for this component to remain informative, but let the updates be handled elsewhere.

> Create React App doesn't use any existing HMR client either, but [implements its own one](https://github.com/facebook/create-react-app/blob/master/packages/react-dev-utils/webpackHotDevClient.js).

I think the best approach will be something along these lines. We can fork `webpack-hot-middleware` publish application status events. This should give us the missing piece of information we need to get the monitor implemented accurately.

> Another way is to keep `webpack-hot-middleware/client` as is, but subscribe to its status updates […]
> The status cycles through a `idle` -> `check` -> `ready` -> `apply` -> `idle` loop.

I don't think this is viable because we're still not informed of the result of application, whether reload is required or not.

Reply:
> React component seems like the wrong fit for managing the updates.

They can be two separate modules: HMR client and a React component that just subscribes to the client's stream of updates and shows status.

> We can fork `webpack-hot-middleware` publish application status events.

The client code is not great and hasn't been touched for a few years. Instead of forking it, reimplementing it from scratch is not much harder and the result would be easier to read and understand.

> I don't think this is viable because we're still not informed of the result of application, whether reload is required or not.

If `applyOptions` doesn't include the `ignore{Errored,Declined,Unaccepted}` flags, the `module.hot.apply` call will return a rejected promise. That means the update failed and full reload is needed. The CRA client does exactly that.

Reply:
I believe we're in agreement 🙂

I'm going to put this on hold and come back to it when I have some time to implement the ideas outlined above. If someone else gets to it first, they're welcome to it.